server/src/main/java/org/opensearch/indices/replication/common/ReplicationTarget.java

```diff
@@ -172,4 +182,83 @@ protected void ensureRefCount() {
         }
     }
 
+    public ActionListener<Void> createOrFinishListener(
+        final TransportChannel channel,
+        final String action,
+        final RecoveryTransportRequest request
+    ) {
+        return createOrFinishListener(channel, action, request, nullVal -> TransportResponse.Empty.INSTANCE);
+    }
+
+    public ActionListener<Void> createOrFinishListener(
+        final TransportChannel channel,
+        final String action,
+        final RecoveryTransportRequest request,
+        final CheckedFunction<Void, TransportResponse, Exception> responseFn
+    ) {
+        final ActionListener<TransportResponse> channelListener = new ChannelActionListener<>(channel, action, request);
+        final ActionListener<Void> voidListener = ActionListener.map(channelListener, responseFn);
+
+        final long requestSeqNo = request.requestSeqNo();
+        final ActionListener<Void> listener;
+        if (requestSeqNo != SequenceNumbers.UNASSIGNED_SEQ_NO) {
+            listener = markRequestReceivedAndCreateListener(requestSeqNo, voidListener);
+        } else {
+            listener = voidListener;
+        }
+
+        return listener;
```

Comment:
Doesn't seem like this would ever be `null` ? If so, you can remove all of the null checks from the numerous times `PeerRecoveryTargetService` invokes this method.

Reply:
`markRequestReceivedAndCreateListener` delegates to [ReplicationRequestTracker.markReceivedAndCreateListener](https://github.com/opensearch-project/OpenSearch/blob/a023ad9cbabdde9c2bf1afb18b73f94357bb478a/server/src/main/java/org/opensearch/indices/replication/common/ReplicationRequestTracker.java#L69) which can return null if the unique request id has already been processed.  In this case it completes the voidListener and returns null.  I'll add nullable all the way down so we are aware of why the checks are necessary.